File: c/questions/connect.c
/**
 * Definition for a Node.
 * struct Node {
 *     int val;
 *     struct Node *left;
 *     struct Node *right;
 *     struct Node *next;
 * };
 */

struct Node* connect(struct Node* root) {
    if (!root) {
        return NULL;
    }
    struct Node* nl[6001];
    nl[0] = root;
    int ls = 0;
    int le = 1;
    int ge = 1;
    int cp;
    struct Node* pre;
    struct Node* cn;
    while (ls < ge) {
        pre = NULL;
        for (cp=ls; cp<le; cp++) {
            cn = nl[cp];
            cn->next = pre;
            pre = cn;
            if (cn->right) {
                nl[ge++] = cn->right;
            }
            if (cn->left) {
                nl[ge++] = cn->left;
            }
        }
        ls = le;
        le = ge;
    }
    return root;
}

File: python/questions/connect.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val: int = 0, left: 'Node' = None, right: 'Node' = None, next: 'Node' = None):
        self.val = val
        self.left = left
        self.right = right
        self.next = next
"""

class Solution:
    def connect(self, root: 'Node') -> 'Node':
        if root is None:
            return None
        q = deque()
        q.append(root)
        while len(q):
            nll = len(q)
            for np in range(nll):
                nn = q.popleft()
                if np == nll-1:
                    nn.next = None
                else:
                    nn.next = q[0]
                if nn.left:
                    q.append(nn.left)
                if nn.right:
                    q.append(nn.right)
        return root
